blocks/logic.ts

```diff
@@ -304,100 +314,107 @@ const CONTROLS_IF_MUTATOR_MIXIN = {
     }
     if (this.elseCount_) {
-      container.setAttribute('else', 1);
+      container.setAttribute('else', '1');
     }
     return container;
   },
   /**
    * Parse XML to restore the else-if and else inputs.
    * Backwards compatible serialization implementation.
-   * @param {!Element} xmlElement XML storage element.
-   * @this {Block}
+   *
+   * @param xmlElement XML storage element.
    */
-  domToMutation: function(xmlElement) {
-    this.elseifCount_ = parseInt(xmlElement.getAttribute('elseif'), 10) || 0;
-    this.elseCount_ = parseInt(xmlElement.getAttribute('else'), 10) || 0;
+  domToMutation: function(this: IfBlock, xmlElement: Element) {
+    const elseifAttribute = xmlElement.getAttribute('elseif');
+    if (elseifAttribute === null) {
+      throw new TypeError('xmlElement did not elseif attribute');
+    }
+    this.elseifCount_ = parseInt(elseifAttribute, 10) || 0;
+    this.elseCount_ = parseInt(xmlElement.getAttribute('else')!, 10) || 0;
     this.rebuildShape_();
   },
   /**
    * Returns the state of this block as a JSON serializable object.
-   * @return {?{elseIfCount: (number|undefined), haseElse: (boolean|undefined)}}
-   *     The state of this block, ie the else if count and else state.
+   * @return The state of this block, ie the else if count and else state.
    */
-  saveExtraState: function() {
-    if (!this.elseifCount_ && !this.elseCount_) {
-      return null;
-    }
-    const state = Object.create(null);
-    if (this.elseifCount_) {
-      state['elseIfCount'] = this.elseifCount_;
-    }
-    if (this.elseCount_) {
-      state['hasElse'] = true;
-    }
-    return state;
-  },
+  saveExtraState: function(this: IfBlock): IfExtraState |
+      null {
+        if (!this.elseifCount_ && !this.elseCount_) {
+          return null;
+        }
+        const state = Object.create(null);
+        if (this.elseifCount_) {
+          state['elseIfCount'] = this.elseifCount_;
+        }
+        if (this.elseCount_) {
+          state['hasElse'] = true;
+        }
+        return state;
+      },
   /**
    * Applies the given state to this block.
-   * @param {*} state The state to apply to this block, ie the else if count and
+   *
+   * @param state The state to apply to this block, ie the else if count
+   and
    *     else state.
    */
-  loadExtraState: function(state) {
+  loadExtraState: function(this: IfBlock, state: IfExtraState) {
     this.elseifCount_ = state['elseIfCount'] || 0;
     this.elseCount_ = state['hasElse'] ? 1 : 0;
     this.updateShape_();
   },
   /**
    * Populate the mutator's dialog with this block's components.
-   * @param {!Workspace} workspace Mutator's workspace.
-   * @return {!Block} Root block in mutator.
-   * @this {Block}
+   *
+   * @param workspace Mutator's workspace.
+   * @return Root block in mutator.
    */
-  decompose: function(workspace) {
+  decompose: function(this: IfBlock, workspace: Workspace): Block {
```

Comment:
Should this return a `ContainerBlock` rather than a `Block`?

Reply:
Good idea.  Done.